### ISSUE TITLE
Failure resolving conda envs in parallel

### ISSUE THREAD

**Issue by grst** [nextflow-io/nextflow#1819]
### Expected behavior and actual behavior

When nextflow tries to create conda environments in parallel, that share a package, environment creation fails with 
the following error. 
```
InvalidArchiveError("Error with archive /opt/conda/pkgs/openjdk-10.0.2-h14c3975_1015.tar.bz2.  You probably need to delete and re-download or re-create this file.  Message from libarchive was:\n\nFailed to create dir 'legal'")
```

Expected: Create environments one after each other to avoid the problem. 

### Steps to reproduce the problem
Minimal nextflow example (`test.nf`)

```nextflow
#!/usr/bin/env nextflow

nextflow.enable.dsl = 2

process test1 {
    conda "python=3.7 star rsem fastqc gatk"
    
    script:
    """
    exit 0
    """
}

process test2 {
    conda "python=3.8 star rsem fastqc gatk"
    
    script:
    """
    exit 0
    """
}

workflow { 
    test1()
    test2()
}

```

**Note:** If the packages are already cached by conda, there will be no error. To test it in a container: 

```bash
podman pull continuumio/miniconda3
podman run -it -v .:/test continuumio/miniconda3 /bin/bash
> conda install -y -c conda-forge -c bioconda nextflow
> nextflow run /test/test.nf
```

### Program output 

```console
(base) root@dc3f51d38a99:/# ./test/test.nf 
N E X T F L O W  ~  version 20.10.0
Launching `./test/test.nf` [agitated_cori] - revision: 647d8085a5
[-        ] process > test1 -
[-        ] process > test1 -
[-        ] process > test2 -
Error executing process > 'test1'

Caused by:
  Failed to create Conda environment
  command: conda create --mkdir --yes --quiet --prefix /work/conda/env-54b51c21323e895919a03b9a835682b4 python=3.7 star rsem fastqc gatk
  status : 1
  message:
    InvalidArchiveError("Error with archive /opt/conda/pkgs/gettext-0.19.8.1-h0b5b191_1005.tar.bz2.  You probably need to delete and re-download or re-create this file.  Message from libarchive was:\n\nFailed to create dir 'share'")
```

### Environment 

* Nextflow version: 20.10.0
* Java version: openjdk version "11.0.8-internal" 2020-07-14
* Operating system: Linux
* Bash version: 5.0.18

### Additional context

This problem became more relevant with DSL2 pipelines, where every process has its own environment. 
Several nf-core users have reported this problem by now (e.g. https://github.com/nf-core/rnaseq/issues/517).

**Comment by pditommaso** [comment]
This means that's not enough to synchronize at the level of the environment directory.  It needs a global lock. 

https://github.com/nextflow-io/nextflow/blob/e94e6877c2dc9bed983e873b62e302a08843c891/modules/nextflow/src/main/groovy/nextflow/conda/CondaCache.groovy#L219-L224

**Comment by phue** [comment]
> It needs a global lock.

How would this work?
I saw that the code to pull singularity images first pulls it to a temporary and then moves it to the actual location when it's there, presumably to avoid this kind of problem(?) https://github.com/nextflow-io/nextflow/blob/abd1c9d9afab8548a16961b5dee6ec0dd4ef17ba/modules/nextflow/src/main/groovy/nextflow/container/SingularityCache.groovy#L217-L231

Would it also be a solution to create the conda environment in a temporary folder first?

**Comment by pditommaso** [comment]
I suspect that conda get messed up when two different recipes (that using a different install dir and therefore can be installed in parallel, have one or more stared dependency which creates a conflict).

**Comment by phue** [comment]
Ok now I understand, the problem is with the global conda cache.
(Related conda issue: https://github.com/conda/conda/issues/8870)

**Comment by abhi18av** [comment]
@phue , I've seen some momentum behind a sibling [mamba](https://github.com/mamba-org/mamba) project which can be a drop-in replacement for `conda` and provides faster performance.

> parallel downloading of repository data and package files using multi-threading

Do you think, it might be worth exploring `mamba`?

**Comment by phue** [comment]
+1 on that, I've been pretty impressed by mamba's speed so far.
I'm not sure that everything that is needed here is actually implemented already (e.g. https://github.com/mamba-org/mamba/issues/633)

**Comment by abhi18av** [comment]
Oh man, that's a bummer as far as the integration is concerned 😓

**Comment by grst** [comment]
Mamba is really great, but I'm not even sure if it solves the problem. It still uses a cache for the packages and it might (?) even share the cache with conda. Certainly worth a try though!

**Comment by pditommaso** [comment]
Is mamba cli compatible with conda (I mean the command line options)?

**Comment by grst** [comment]
yes, it's meant as a drop-in replacement. 

only `activate` should still be `conda activate`. 

Plus is something is not implemented in mamba (as mentioned above) it will fall back to the corresponding conda command.

**Comment by pditommaso** [comment]
If anyone wants to try mamba support for NF it's welcome.

**Comment by abhi18av** [comment]
> If anyone wants to try mamba support for NF it's welcome.

There are a couple ways which I can think of to approach this: 

1. Clearly, `mamba` is based on `conda` and is heavily invested in the protocols and CLI options setup by the `conda` team - we could simply add another option to the `conda` directive called `condaBin` as shown below

```nextflow

process COW_SAY_MOO {
conda 'cowsay', condaBin: "/path/to/conda-or-mamba"

script: 
"""
cowsay "Moo"
"""
}


```

This imght be _safer_ since we don't know how `mamba` project would pan out in the long term and therefore it makes sense to keep the changes minimal.

2. Keeping in mind that `mamba` is not a standalone project, it has other other aspects like a package builder `boa` and package server `quetz` (more exiciting) - we could exploit this ability later on to enhance the Nextflow and family.

With this perspective, it makes sense to introduce a new `mamba` directive with associated options like `packageServerEndpoint` etc

I might have missed out something here, please feel free to highlight and suggest improvements :)

**Comment by grst** [comment]
I think it would be nice to be able to configure `condaBin` per project, and not per process. I think this would already be useful for cases where `conda` is not in `PATH`. 

While I'm all for `mamba` for its speed benefit, it doesn't solve the particular problem raised in this thread. Mamba shares the package cache with conda and I just tried running the following two mamba commands in parallel and it also failed with a race condition: 

1. clean conda cache
```
conda clean -af
```

2. run in two terminal windows in parallel
```bash
mamba create -n test-mamba2 python=3.8 star rsem fastqc gatk
```

```bash
mamba create -n test-mamba1 python=3.7 star rsem fastqc gatk
```

3. Error message: 
```console
terminate called after throwing an instance of 'std::runtime_error'
  what():  truncated bzip2 input
zsh: abort (core dumped)  mamba create -n test-mamba2 python=3.8 star rsem fastqc gatk
```

**Comment by abhi18av** [comment]
I've had discussion with the `mamba` author and this ability would be available soon in `mamba` 🤞 

Cross-referencing here 

https://github.com/mamba-org/mamba/issues/706

**Comment by pditommaso** [comment]
Since mamba is solving this and looks much faster than conda, does anybody want to try to contribute the support for it?

**Comment by abhi18av** [comment]
If that's okay I'd like to take this forward and would be happy to collaborate with anyone for the review/code and testing.

**Comment by drpatelh** [comment]
Getting the same error as reported in the original comment whilst testing the latest `dev` version of nf-core/rnaseq before release. The problem persists after multiple `-resume`.

```console
$ nextflow pull nf-core/rnaseq
$ nextflow run nf-core/rnaseq -profile test,conda -r dev

Caused by:
  Failed to create Conda environment
  command: conda create --mkdir --yes --quiet --prefix rnaseq/test_conda/work/conda/env-7e075e748f6bfc7634241fa77a653e1e bioconda::trim-galore=0.6.6
  status : 1
  message:
    CondaMultiError: InvalidArchiveError(u"Error with archive conda/pkgs/openjdk-11.0.9.1-h5cc2fde_1.tar.bz2.  You probably need to delete and re-download or re-create this file.  Message from libarchive was:\n\nFailed to create dir 'lib'",)
```

**Comment by grst** [comment]
Still getting this error with mamba on our CI: 

```
Caused by:
  Failed to create Conda environment
  command: mamba create --mkdir --yes --quiet --prefix /home/runner/pytest_workflow_hwxk5qaf/deseq2/work/conda/env-8267b5afea7686b0592f9121d90bd5b7 conda-forge::r-base=4.1.2 conda-forge::r-docopt=0.7.1 conda-forge::r-conflicted=1.1.0 conda-forge::r-readr=2.1.1 conda-forge::dplyr=1.0.7 bioconda::bioconductor-enhancedvolcano=1.12.0
  status : 1
  message:
    ERROR   Selected channel specific (or force-reinstall) job, but package is not available from channel. Solve job will fail.
    
    LockError: 
    LOCKERROR: It looks like conda is already doing something.
    The lock ['/home/runner/conda_pkgs_dir.pid6646.conda_lock'] was found. Wait for it to finish before continuing.
    If you are sure that conda is not running, remove it and try again.
    You can also use: $ conda clean --lock
```

At least the error message has improved, but the pipeline still can't complete successfully.

**Comment by jpfeuffer** [comment]
I am having the same problem. Unfortunately, not even queueSize=1 solves the problem. Does nextflow try to setup conda environments before the actual steps are performed?

**Comment by lindenb** [comment]
Hi I got the same problem today using mamba:


```
NXF_CONDA_CACHEDIR=/CONDAS/users/lindenbaum-p \                                                                
nextflow run -c confs/main.cfg -profile sge -resume -work-dir 'path/to/work'  \        
        --samplesheet  samplesheet.txt  main.nf

N E X T F L O W  ~  version 23.04.1
Launching `main.nf` [shrivelled_tesla] DSL2 - revision: 16a6447af6
Creating env using mamba: bioconda:samtools=1.18 [cache /CONDAS/users/lindenbaum-p/env-2bb3f51fb7a0ae29ccda431ea048dec6]
Creating env using mamba: bioconda:fastqc=0.12.1 [cache /CONDAS/users/lindenbaum-p/env-5e19dbc025988a06b84475e33b8b6f59]
Creating env using mamba: bioconda:trim-galore=0.6.10 [cache /CONDAS/users/lindenbaum-p/env-4b72977e73b345465fc899715903378a]
ERROR ~ Error executing process > 'ATACSEQ:TRIM_GALORE:FASTQC (1)'
                                                                                                               
Caused by:                                                                                                     
  Failed to create Conda environment       
  command: mamba create --mkdir --yes --quiet --prefix /CONDAS/users/lindenbaum-p/env-5e19dbc025988a06b84475e33b8b6f59 bioconda:fastqc=0.12.1
  status : 1                                  
  message:                                  
    error    libmamba Could not set lock (Resource temporarily unavailable)
    warning  libmamba Cannot lock '/CONDAS/users/lindenbaum-p/packages/cache'                       
        Waiting for other mamba process to finish                                                              
 
```

**Comment by bentsherman** [comment]
This problem was fixed in 23.09.0-edge

**Comment by pditommaso** [comment]
Still local conda builds are slow and fragile. Pierre, you should really give a try Conda[-based containers delivered on-demand](https://www.nextflow.io/docs/edge/wave.html#build-conda-based-containers), it works like a charm!
